utilities/backup/backup_engine.cc

```diff
@@ -2079,14 +2077,9 @@ IOStatus BackupEngineImpl::CopyOrCreateFile(
     io_s = dest_writer->Append(data);
 
     if (rate_limiter != nullptr) {
-      if (!src.empty()) {
-        rate_limiter->Request(data.size(), Env::IO_LOW, nullptr /* stats */,
```

Comment:
I believe in the case of `if (!src.empty())` we don't need to loop as we still have `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` [above](https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6L2030-L2032)  AND in the case of `!src.empty()`, we are writing what we read and the constrained `buf_size` still takes effect as it is passed down in `io_s = src_reader->Read(buffer_to_read, &data, buf.get(), Env::IO_LOW)` [here](https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6R2056). So the `buffer_to_read` >= data.size() that we are going to write.


Unlike the else case  (i.e, `src.empty() ==  true`) where we are writing whatever is passed into this function (i.e, `data = contents;` in https://github.com/facebook/rocksdb/pull/9973/files?show-viewed-files=true&file-filters%5B%5D=#diff-4d85a375a1fe9d14d0e5cfa18bf19729e614a4f3edf73ca59b0e7bde0b1e79c6L2061) so the constrained buffer_to_read does not take effect.


However, using `LoopRateLimitRequestHelper` for both case like the change here should be fine as the `LoopRateLimitRequestHelper` just iterate once for the former case.

Reply:
[No action needed from author and just a note]

I think having `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` actually limits the exercise of the newly added lower-level rate-limiting in SequentialFileReader as the single read `io_s = src_reader->Read(buffer_to_read...` is already within burst.  

But `size_t buf_size = rate_limiter ? static_cast<size_t>(rate_limiter->GetSingleBurstBytes()): kDefaultCopyFileBufferSize;` is still needed as we haven't migrated the write rate-limiting. Remove it while not migrating write will leads to `io_s = dest_writer->Append(data);` [here](https://github.com/facebook/rocksdb/blob/0d8433c5853e468975382b05b92cdd72379c64c8/utilities/backup/backup_engine.cc#L2077) actually write-more-than-burst before rate-limit.